### PR TITLE
fixes #1167 - User specific data now shown by githubprofilescraper se…

### LINKED ISSUE #1167: User specific data not shown by githubprofilescraper service
- Issue type: Bug report

### Short description
githubprofilescraper service is not showing user specific data like followers_url, number of followers, following_url, number of people following, starred_url, number of starred repos.

git

### PR DESCRIPTION
**Pull request by djmgit**
…rvice

Fixes issue #1167.githubprofilescraper service now displays starred_url,
number of starred repos,followers_url, number of followers, following_url,
number of people following for a particuler user.

### Short description
I have:
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message 
- [x] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

user

Test link: 
http://127.0.0.1:9000/api/githubprofilescraper.json?profile=djmgit

updated results:
starred_url : https://api.github.com/users/djmgit/starred
followers_url : https://api.github.com/users/djmgit/followers
following_url : https://api.github.com/users/djmgit/following

### CODE REVIEW

File: src/org/loklak/api/search/GithubProfileScraper.java [hunk 1]
```diff
@@ -149,24 +149,24 @@ public static SusiThought scrapeGithub(String profile) {
...
+		if (html.getElementsByAttributeValueContaining("class", "user-profile-nav").size() != 0) {
 
-			String followersUrl = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(0).attr("href");
-			githubProfile.put("followers_url", "https://github.com" + followersUrl);
+			String starredUrl = "https://api.github.com/users/" + profile + "/stars";
+			githubProfile.put("starred_url", starredUrl);
 
-			String followers = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(0).tagName("strong").text();
-			githubProfile.put("followers", followers);
+			String starred = html.getElementsByAttributeValue("class", "Counter").get(1).text();
```
Comment (singhpratyush) [inline src/org/loklak/api/search/GithubProfileScraper.java]: I think this should be a numeric type variable.

**Comment by singhpratyush** [comment]
Please mark corresponding checkboxes in PR description and give updated results for the issue there (try putting text as it becomes difficult to copy from an image :smile:).

**Comment by kavithaenair** [comment]
@djmgit Screenshot of output json please.

**Comment by djmgit** [comment]
@singhpratyush @kavithaenair did the changes, please review :)

**Comment by djmgit** [comment]
@singhpratyush yes I will be adding tests, actually some more work needs to be done on githubprofilescraper. Presently its showing some unnecessary fields for organizations. I am presently working on it.Once its solved I will add the corresponding tests and create another issue

File: src/org/loklak/api/search/GithubProfileScraper.java [hunk 2]
```diff
@@ -149,24 +149,24 @@ public static SusiThought scrapeGithub(String profile) {
...
+			githubProfile.put("starred_url", starredUrl);
 
-			String followers = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(0).tagName("strong").text();
-			githubProfile.put("followers", followers);
+			int starred = Integer.parseInt(html.getElementsByAttributeValue("class", "Counter").get(1).text());
+			githubProfile.put("starred", starred);
 
-			String starredUrl = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(1).attr("href");
-			githubProfile.put("starred_url", "https://github.com" + starredUrl);
+			String followersUrl = "https://api.github.com/users/" + profile + "/followers";
```
Comment (sudheesh001) [inline src/org/loklak/api/search/GithubProfileScraper.java]: Though I like this, the problem with this content is the fact that there are hardcoded URL bases which can easily be refactored to.

`String GITHUB_API_BASE = 'https://api.github.com/users/'`

Similarly the endpoints (of course you can come up with better variable names)
```java
String STARRED_ENDPOINT = "/starred";
String FOLLOWERS_ENDPOINT = "/followers";
```

File: src/org/loklak/api/search/GithubProfileScraper.java [hunk 3]
```diff
@@ -149,24 +149,24 @@ public static SusiThought scrapeGithub(String profile) {
...
+			githubProfile.put("followers", followers);
 
-			String followingUrl = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(2).attr("href");
-			githubProfile.put("following_url", "https://github.com" + followingUrl);
+			String followingUrl = "https://api.github.com/users/" + profile + "/following";
+			githubProfile.put("following_url", followingUrl);
 
-			String following = html.getElementsByAttributeValueContaining("class", "vcard-stat").get(2).tagName("strong").text();
+			int following = Integer.parseInt(html.getElementsByAttributeValue("class", "Counter").get(3).text());
 			githubProfile.put("following", following);
```
Comment (sudheesh001) [inline src/org/loklak/api/search/GithubProfileScraper.java]: Though I agree that all of this information would be useful, Why not create multiple requests from here for the API endpoints encountered and store that information too? One definitely needs to be careful about the request limit here for github's public APIs.

Do think about how as a developer you might interact with this API, What are the other content that you'd want to pass and check?
```
profile: - Users profile to scrape and store
terms: [ followers, following, starred, repos ]
```
If a user makes a request for profile, the current state of the scraper sounds good but if there are more additions to it in terms, maybe the user wants to know the `followers` of a specific user in `profile`, similarly with the other parameters.

**Comment by djmgit** [comment]
Changes I made:
1. Re-factored  the code and separated code scraping individual data from organization data.
2. Made requests for the various endpoints encountered and stored the data rather than just storing the 
   corresponding links.
3. test url : http://127.0.0.1:9000/api/githubprofilescraper.json?q=djmgit will fetch all the data
4. User can specify the data he/she wants by including specific fields in terms list.
    For example if user wants only following and followers of a user then the following url can be
    used : http: //127.0.0.1:9000/api/githubprofilescraper.json?profile=djmgit&terms=followers,following
    In this way githubprofilescraper will not scrape the data which is not required and respond faster

@singhpratyush @kavithaenair @sudheesh001 open for review

File: src/org/loklak/api/search/GithubProfileScraper.java [hunk 4]
```diff
@@ -66,53 +70,42 @@ public JSONObject getDefaultPermissions(BaseUserRole baseUserRole) {
 	public JSONObject serviceImpl(Query call, HttpServletResponse response, Authorization rights, JSONObjectWithDefault permissions)
 			throws APIException {
 		String profile = call.get("profile", "");
+		String termsS = call.get("terms", "");
```
Comment (sudheesh001) [inline src/org/loklak/api/search/GithubProfileScraper.java]: This is a weird name to give and a confusing `sS`. Is there an issue if we call it `terms`? That seems better.

Comment (djmgit) [inline src/org/loklak/api/search/GithubProfileScraper.java]: @sudheesh001 Actually just after this I have declared a Set named terms, so if I change termsS to terms, java is showing an exception : variable terms is aldready defined at  ```Set terms = null ```.

**Comment by djmgit** [comment]
@sudheesh001 changed termsS to termsParam

**Comment by SarishiNoHara** [comment]
Hi @djmgit , I wonder how to use the GithubProfileScraper. It returns server error when I enter these links.

**Comment by djmgit** [comment]
@SarishiNoHara try this :
http://127.0.0.1:9000/api/githubprofilescraper.json?profile=djmgit
http://loklak.org/api/githubprofilescraper.json?profile=djmgit&terms=followers,following

You need to set the profile parameter to the profile name you are looking for and terms parameter to specify what information you need to scrape. If you do not specify terms parameter then scraper will return all the data.

**Comment by SarishiNoHara** [comment]
Yeah, it works, but it doesn't return all the data if I don't add any parameter.
[http://loklak.org/api/githubprofilescraper.json](http://loklak.org/api/githubprofilescraper.json)
[http://loklak.org/api/githubprofilescraper.json?profile=](http://loklak.org/api/githubprofilescraper.json?profile=)

**Comment by djmgit** [comment]
@SarishiNoHara you need to call the service on your local server now, as the changes have not yet been made on loklak.org server

**Pull request merged**